pkg/net/network.go

```diff
@@ -234,77 +136,35 @@ func (n *network) Listen(ctx context.Context, address string) (chan *Connection,
 	}
 
 	logger.Info("Started listening", zap.Strings("addresses", addresses))
-	n.addAddress(addresses...)
-	n.addAddress(n.relayAddresses...)
+	n.local.AddAddress(addresses...)
+	n.local.AddAddress(n.relayAddresses...)
 
 	cconn := make(chan *Connection, 10)
 	go func() {
 
 		for {
 			tcpConn, err := tcpListener.Accept()
 			if err != nil {
-				log.DefaultLogger.Warn("could not accept connection", zap.Error(err))
-				// TODO close conn?
-				return
+				log.DefaultLogger.Warn(
+					"could not accept connection", zap.Error(err))
+				continue
 			}
 
 			conn := &Connection{
 				Conn:          tcpConn,
 				RemotePeerKey: nil,
+				IsIncoming:    true,
 			}
 
-			synObj, err := Read(conn)
-			if err != nil {
-				log.DefaultLogger.Warn("waiting for syn failed", zap.Error(err))
-				// TODO close conn?
-				continue
-			}
-
-			syn := &HandshakeSyn{}
-			if err := syn.FromObject(synObj); err != nil {
-				log.DefaultLogger.Warn("could not convert obj to syn")
-				// TODO close conn?
-				continue
-			}
-
-			// store the remote peer
-			conn.RemotePeerKey = syn.PeerInfo.SignerKey
-			n.discoverer.Add(syn.PeerInfo)
-
-			synAck := &HandshakeSynAck{
-				Nonce:    syn.Nonce,
-				PeerInfo: n.GetPeerInfo(),
-			}
-			sao := synAck.ToObject()
-			if err := crypto.Sign(sao, n.key); err != nil {
-				log.DefaultLogger.Warn("could not sign for syn ack object", zap.Error(err))
-				// TODO close conn?
-				continue
-			}
-			if err := Write(sao, conn); err != nil {
-				log.DefaultLogger.Warn("sending for syn-ack failed", zap.Error(err))
-				// TODO close conn?
-				continue
-			}
-
-			ackObj, err := Read(conn)
-			if err != nil {
-				log.DefaultLogger.Warn("waiting for ack failed", zap.Error(err))
-				// TODO close conn?
-				continue
-			}
+			for _, mh := range n.middleware {
+				conn, err = mh(ctx, conn)
+				if err != nil {
+					log.DefaultLogger.Error(
+						"middleware failure", zap.Error((err)))
 
-			ack := &HandshakeAck{}
-			if err := ack.FromObject(ackObj); err != nil {
-				// TODO close conn?
-				log.DefaultLogger.Warn("could not convert obj to syn ack")
-				continue
-			}
-
-			if ack.Nonce != syn.Nonce {
-				log.DefaultLogger.Warn("validating syn to ack nonce failed")
-				// TODO close conn?
-				continue
+					tcpConn.Close()
+					break
```

Comment:
I think here in case of error you are breaking out of the middleware loop and pushing a nil connection to the channel.

You are also closing the top connection instead of the wrapped connection (not applicable right now as we only have one middleware), so not really important but can we add a todo to close the last known wrapped connection instead?